### PR TITLE
Instructor: view results: give a 'Edit view' button #7297

### LINKED ISSUE #7297: Instructor: view results: give a 'Edit view' button
Current:
The page is reloaded whenever a setting of the view is changed:
image

Suggested:
Make the settings un-editable.  Provide an `Edit view` button that shows the settings in a modal window.
image

Justification:
Reloading page for each settings change is a waste of resources (and can be irritating to the user) because the user might want to change multiple settings.

### PR DESCRIPTION
**Pull request by LimeTheCoder**
Fixes #7297 

editdemonstration

### CODE REVIEW

**Comment by LimeTheCoder** [comment]
Need review

**Comment by damithc** [comment]
> Using an extra modal to edit the settings is not very aesthetically pleasing, as the original greyed-out checkboxes and dropdowns remain.
> 
> Perhaps what could be done is to just remove the AJAX requests made when the checkboxes and buttons are changed. Then a single AJAX request can be made when the user clicks "Edit View", based on the state of the checkboxes and buttons at the time of the click.
> 
> This way, the original interface can be kept without the annoyance of a page reload every time something is changed.

I agree the modal is a bit 'overdoing' but I kind of favor it in this situation because existing users will immediately recognize it as a new way to update the view (I worry that existing users will change settings and wait for the page to load without realizing that the new UI requires one more action for the page to refresh.). Also, the modal is a clear indication that the user is in 'edit mode' and it is also easy to cancel an edit and go back to the original view.

**Comment by thenaesh** [comment]
@damithc In that case, perhaps the original view could be modified a bit to simply state what the current view is, rather than making it look like an editable box? Then, when the "Edit View" button is clicked, the user will naturally expect a modal to show up.

My concern now is that the original interface appears to be editable (or rather, it looks like clicking the "Edit View" button will make the existing fields editable). So it may be a bit jarring when the user clicks on it and they get a modal instead.

**Comment by damithc** [comment]
> @damithc In that case, perhaps the original view could be modified a bit to simply state what the current view is, rather than making it look like an editable box? Then, when the "Edit View" button is clicked, the user will naturally expect a modal to show up.
> 
> My concern now is that the original interface appears to be editable (or rather, it looks like clicking the "Edit View" button will make the existing fields editable). So it may be a bit jarring when the user clicks on it and they get a modal instead.

Good point. Yes, we should make it NOT look like a form.

**Comment by LimeTheCoder** [comment]
Updated. Original view doesn't look like a form.

edit2

**Comment by damithc** [comment]
Good work. 
I prefer to show all three tick-box options in the page though as it is more informative to the user. 
Perhaps we can use ~~strike off~~ to show which ones are applied?

**Comment by LimeTheCoder** [comment]
Updated
peek 2017-07-04 00-54

**Comment by LimeTheCoder** [comment]
Updated.

**Comment by damithc** [comment]
Good work @LimeTheCoder 
@whipermr5 @chowyb can one of you do the 2nd review for this? I don't have much experience with this part of the code.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+<div id="editModal" class="modal fade" role="dialog">
+  <div class="modal-dialog">
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal">&times;</button>
+        <h4 class="modal-title">Edit settings</h4>
+      </div>
+      <div class="modal-body">
+        <form role="form" method="post" action="${filterPanel.resultsLink}">
```
Comment (thenaesh) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: in HTML5, the ARIA attribute `role="form"` is already [implicitly set](https://www.w3.org/TR/html-aria/#dfn-implicit-aria-semantics), so it doesn't have to be mentioned again.

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java [hunk 2]
```diff
@@ -75,46 +75,90 @@ public boolean isCorrectPage(String courseId, String feedbackSessionName) {
         return isCorrectCourseId && isCorrectFeedbackSessionName && containsExpectedPageContents();
     }
 
+    public void displayEditSettingsWindow() {
+        WebElement editBtn = browser.driver.findElement(By.id("editBtn"));
+        editBtn.click();
+        WebDriverWait wait = new WebDriverWait(browser.driver, TestProperties.TEST_TIMEOUT);
+        wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("editModal")));
```
Comment (whipermr5) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackResultsPage.java]: Extract this into an `AppPage` method `waitForElementVisibility`?

**Comment by LimeTheCoder** [comment]
Hi, @chowyb. Need your review here.

**Comment by LimeTheCoder** [comment]
Hi, @whipermr5 . Need your review)

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+          <input type="hidden" name="<%=Const.ParamsNames.FEEDBACK_SESSION_NAME%>"
+                 value="${filterPanel.feedbackSessionName}">
+          <input type="hidden" name="<%=Const.ParamsNames.COURSE_ID%>"
+                 value="${filterPanel.courseId}">
+          <input type="hidden" name="<%=Const.ParamsNames.USER_ID%>"
+                 value="${data.account.googleId}">
+          <hr>
+          <div class="row">
+            <div class="col-md-2">
+              <button id="submitBtn" type="submit" class="btn btn-primary">Submit</button>
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Can the button label be `Change View`? Clearer that clicking it will load the page with a new view.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            </div>
+            <div class="col-md-offset-8 col-md-2">
+              <button type="button" class="btn btn-danger" data-dismiss="modal">Close</button>
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Can the button label be `Cancel`? Clearer that clicking it won't do anything.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+<%@ tag import="teammates.common.util.Const" %>
+<%@ taglib uri="http://java.sun.com/jsp/jstl/functions" prefix="fn" %>
+<%@ attribute name="filterPanel" type="teammates.ui.template.InstructorFeedbackResultsFilterPanel" required="true" %>
+
+<div id="editModal" class="modal fade" role="dialog">
+  <div class="modal-dialog">
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal">&times;</button>
+        <h4 class="modal-title">Edit settings</h4>
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Can the title be `Edit View`? The word `settings` is pretty vague in this context.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterPanel.tag [hunk 6]
```diff
@@ -1,121 +1,97 @@
...
-                <a class="btn btn-default btn-xs pull-right" id="collapse-panels-button" data-toggle="tooltip" title="Expand all panels. You can also click on the panel heading to toggle each one individually.">
-                    Expand All ${filterPanel.sortType == 'question' ? 'Questions' : 'Sections'}
-                </a>
-            </c:otherwise>
-        </c:choose>
+    <div class="row">
+      <div class="col-md-offset-10 col-md-2">
+        <button id="editBtn" type="button" class="btn btn-primary btn-block"
+                data-toggle="modal" data-target="#editModal">
+          Edit settings
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterPanel.tag]: Rename to `Edit View` here too.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+          </div>
+
+          <input type="hidden" name="<%=Const.ParamsNames.FEEDBACK_SESSION_NAME%>"
+                 value="${filterPanel.feedbackSessionName}">
+          <input type="hidden" name="<%=Const.ParamsNames.COURSE_ID%>"
+                 value="${filterPanel.courseId}">
+          <input type="hidden" name="<%=Const.ParamsNames.USER_ID%>"
+                 value="${data.account.googleId}">
+          <hr>
+          <div class="row">
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Delete the row and move the buttons up to this level instead. Add the `pull-right` class to the cancel button. This will allow it to look good on all screen resolutions.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditButton.tag [hunk 8]
```diff
@@ -0,0 +1,21 @@
+<%@ tag description="instructorFeedbackResultsTop - Filter Panel Edit Button" %>
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditButton.tag]: Just highlighting that this will conflict with similar changes made in #7665. Do you mind waiting for that one to be merged first? I will expedite it.

Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditButton.tag]: It'd be easier to revert this change first (i.e. it's okay if the expand/collapse button stays inlined within the filter panel for now) for a cleaner merge with #7665.

Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditButton.tag]: Also, `filterEditButton` is a misleading name, as this is actually the expand/collapse button.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 9]
```diff
@@ -0,0 +1,120 @@
+<%@ tag description="instructorFeedbackResultsTop - Filter Panel Edit Modal" %>
+<%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>
+<%@ tag import="teammates.common.util.Const" %>
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Move this after the the `taglib` line below.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterPanel.tag [hunk 10]
```diff
@@ -1,121 +1,97 @@
 <%@ tag description="instructorFeedbackResultsTop - Filter Panel" %>
 <%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>
+<%@ taglib tagdir="/WEB-INF/tags/instructor/results" prefix="r" %>
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterPanel.tag]: Same here.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag (hunk 7 above)
Comment (LimeTheCoder) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Here we also need to manage the size of the button to be consistent with buttons above. With your approach it's will be hard to do so. Also, here used grid system from bootstrap, so the button will be responsive and look good regarding screen resolution.

Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: I encourage you to actually try it out on smaller resolutions (resize the window). The hardcoded offset 8 leads to problems for `xs` screens - the cancel button wraps to the next line. The amount of space in between buttons should be allowed to vary freely, and not hardcoded to 8.

Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Also, just adding an offset 8 isn't actually aligning the button to the right. The button is aligned to the left of the 10th column (and even then, only for resolutions `md` and higher).

Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: The size of the button depends on the font size of the button label, which varies across browsers. It should not be managed using a grid when all we want is to align one to the left and one to the right.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 11]
```diff
@@ -0,0 +1,120 @@
+<%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>
+<%@ taglib uri="http://java.sun.com/jsp/jstl/functions" prefix="fn" %>
+<%@ tag description="instructorFeedbackResultsTop - Filter Panel Edit Modal" %>
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: The tag description should still be the very first line. I meant that the tag import should not come between core and functions. Just move the description to the first line and all will be fine.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterPanel.tag [hunk 12]
```diff
@@ -1,103 +1,95 @@
-<%@ tag description="instructorFeedbackResultsTop - Filter Panel" %>
 <%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>
+<%@ taglib tagdir="/WEB-INF/tags/instructor/results" prefix="r" %>
 <%@ taglib uri="http://java.sun.com/jsp/jstl/functions" prefix="fn" %>
+<%@ tag description="instructorFeedbackResultsTop - Filter Panel" %>
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterPanel.tag]: Same here - move the description back to the top. It is the results (prefix r) that should be moved below functions (prefix fn).

Comment (LimeTheCoder) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterPanel.tag]: It's my fault, sorry;) Good you please review, before I'll run godmode tests.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+                 value="${filterPanel.feedbackSessionName}">
+          <input type="hidden" name="<%=Const.ParamsNames.COURSE_ID%>"
+                 value="${filterPanel.courseId}">
+          <input type="hidden" name="<%=Const.ParamsNames.USER_ID%>"
+                 value="${data.account.googleId}">
+
+          <div class="modal-footer">
+            <button id="submitBtn" type="submit" class="btn btn-primary pull-left">Change View</button>
+            <button type="button" class="btn btn-danger pull-right" data-dismiss="modal">Cancel</button>
+          </div>
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Just realised another problem: `modal-footer` should not be inside `modal-body`, but directly inside `modal-content` (take a look at how the other modals are done e.g. the bootbox one). Placing `modal-footer` inside `modal-body` causes the buttons not to align to the content.

Also, just realised that `modal-footer` already aligns content to the right, so `pull-right` can be dropped entirely. For consistency with the confirm publish/unpublish modal, the cancel button should be on the left of the confirm button, with both buttons together side-by-side on the bottom right corner of the modal. Hence, drop `pull-left` too.

Summary: place this HTML directly inside the `modal-content`, after `modal-body` (i.e., after line 113 in the current version):
```HTML
<div class="modal-footer">
  <button type="button" class="btn btn-danger pull-right" data-dismiss="modal">Cancel</button>
  <button id="submitBtn" type="submit" class="btn btn-primary pull-left">Change View</button>
</div>
```

The result should look like this:
screen shot 2017-07-27 at 6 33 47 pm

Comment (damithc) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Should we use red color for the cancel button? May be reserve red for destructive actions such as `delete` and use a different color for cancel? Does bootstrap has a recommended style for cancel buttons?

Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Agree; the cancel button should be uncoloured for consistency with our other modals:
screen shot 2017-07-27 at 7 45 42 pm

Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: I also recommend adding `alert-info` to the `modal-header` to add a bluish tint and make it look less plain:
screen shot 2017-07-27 at 7 51 47 pm

Comment (LimeTheCoder) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Thanks, now modal looks like this:

image

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+<div id="editModal" class="modal fade" role="dialog">
+  <div class="modal-dialog">
+    <form method="post" action="${filterPanel.resultsLink}">
+      <div class="modal-content">
+        <div class="modal-header alert-info">
+          <button type="button" class="close" data-dismiss="modal">&times;</button>
+          <h4 class="modal-title">Edit View</h4>
+        </div>
+        <div class="modal-body">
+            <div class="row">
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: The indentation level is wrong here.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+                   value="${filterPanel.feedbackSessionName}">
+            <input type="hidden" name="<%=Const.ParamsNames.COURSE_ID%>"
+                   value="${filterPanel.courseId}">
+            <input type="hidden" name="<%=Const.ParamsNames.USER_ID%>"
+                   value="${data.account.googleId}">
+        </div>
+
+        <div class="modal-footer">
+          <button id="submitBtn" type="submit" class="btn btn-primary pull-right margin-left-7px">Change View</button>
+          <button type="button" class="btn btn-default pull-right" data-dismiss="modal">Cancel</button>
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: `pull-right` can be dropped as `modal-footer` already aligns content to the right. Also, no need for `margin-left`.

File: src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+                 value="${filterPanel.feedbackSessionName}">
+          <input type="hidden" name="<%=Const.ParamsNames.COURSE_ID%>"
+                 value="${filterPanel.courseId}">
+          <input type="hidden" name="<%=Const.ParamsNames.USER_ID%>"
+                 value="${data.account.googleId}">
+        </div>
+
+        <div class="modal-footer">
+          <button type="button" class="btn btn-default" data-dismiss="modal">Cancel</button>
+          <button id="submitBtn" type="submit" class="btn btn-primary margin-left-7px">Change View</button>
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/filterEditModal.tag]: Can drop the `margin-left-7px`; it doesn't do anything anyway as it's overwritten by bootstrap's `.modal-footer .btn+.btn`.

**Comment by whipermr5** [comment]
@chowyb You want to take a quick look too?

**Comment by whipermr5** [comment]
@LimeTheCoder Build is failing?

**Comment by LimeTheCoder** [comment]
Fixed failed tests
